tests/http_api_test.go

```diff
@@ -0,0 +1,148 @@
+package tests
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+type SomeJSONStruct struct {
+	SomeField string `json:"someField"`
+}
+
+func (s *clientIntegrationSuite) TestHTTPAPIBasics() {
+	// Create basic workflow that can answer queries, get signals, etc
+	workflowFn := func(ctx workflow.Context, arg *SomeJSONStruct) (*SomeJSONStruct, error) {
+		// Query that just returns query arg
+		err := workflow.SetQueryHandler(ctx, "some-query", func(queryArg *SomeJSONStruct) (*SomeJSONStruct, error) {
+			return queryArg, nil
+		})
+		if err != nil {
+			return nil, err
+		}
+		// Wait for signal to complete
+		var done bool
+		sel := workflow.NewSelector(ctx)
+		sel.AddReceive(workflow.GetSignalChannel(ctx, "some-signal"), func(ch workflow.ReceiveChannel, _ bool) {
+			var signalArg SomeJSONStruct
+			ch.Receive(ctx, &signalArg)
+			if signalArg.SomeField != "signal-arg" {
+				panic("invalid signal arg")
+			}
+			done = true
+		})
+		for !done {
+			sel.Select(ctx)
+		}
+		return arg, nil
+	}
+	s.worker.RegisterWorkflowWithOptions(workflowFn, workflow.RegisterOptions{Name: "http-basic-workflow"})
+
+	// Start
+	workflowID := s.randomizeStr("wf")
+	_, respBody := s.httpPost(http.StatusOK, "/api/v1/namespaces/"+s.namespace+"/workflows/"+workflowID, `{
+		"workflowType": { "name": "http-basic-workflow" },
+		"taskQueue": { "name": "`+s.taskQueue+`" },
+		"input": [{ "someField": "workflow-arg" }],
+		"requestId": "`+s.randomizeStr("req")+`"
+	}`)
+	var startResp struct {
+		RunID string `json:"runId"`
+	}
+	s.Require().NoError(json.Unmarshal(respBody, &startResp))
+
+	// Confirm already exists error with details and proper code
+	_, respBody = s.httpPost(http.StatusConflict, "/api/v1/namespaces/"+s.namespace+"/workflows/"+workflowID, `{
+		"workflowType": { "name": "http-basic-workflow" },
+		"taskQueue": { "name": "`+s.taskQueue+`" },
+		"input": [{ "someField": "workflow-arg" }],
+		"requestId": "`+s.randomizeStr("req")+`"
+	}`)
+	var errResp struct {
+		Message string `json:"message"`
+		Details []struct {
+			RunID string `json:"runId"`
+		} `json:"details"`
+	}
+	s.Require().NoError(json.Unmarshal(respBody, &errResp))
+	s.Require().Contains(errResp.Message, "already running")
+	s.Require().Equal(startResp.RunID, errResp.Details[0].RunID)
+
+	// Query
+	_, respBody = s.httpPost(
+		http.StatusOK,
+		"/api/v1/namespaces/"+s.namespace+"/workflows/"+workflowID+"/query/some-query",
+		`{ "query": { "queryArgs": [{ "someField": "query-arg" }] } }`,
```

Comment:
Notice the inconsistencies in our API come to the forefront here. What is `query.queryArgs` here is `input` for signal/workflow.

Reply:
These are new APIs, though. Why not address this?

Reply:
You cannot address this reasonably with gRPC gateway. Our HTTP API, as a rule and as a byproduct of our approach, exposes the RPCs we have made, it does not alter then. We chose to do this in our API, gRPC and HTTP users have to deal with it. This is covered a bit better in the proposal: https://github.com/temporalio/proposals/blob/master/api/http-api.md.

If we want to fix our RPCs to be more consistent, HTTP and gRPC will benefit. If not, HTTP and gRPC calls will be inconsistent across calls (but consistent across protocols).